### ISSUE TITLE
[1.4.2] Crash with PathToDirt

### ISSUE THREAD

**Issue by Crystal-Spider** [MrUnknown404/SlabGen#2]
**Describe the bug**
When using this mod in combination with [PathToDirt](https://github.com/Insane96/PathToDirt), trying to make a grass slab into a path slab makes the game crash.

**To Reproduce**
1. Create and open a new world.
2. Right-click on a grass slab with any shovel.
3. See the crash

**Expected behavior**
The grass, dirt and path slabs should all behave as normal grass, dirt and path blocks do with PathToDirt.

**Additional context**
Minecraft: 1.18.2
Forge: 40.1.57
PathToDirt: 1.4.2
Existing issue on PathToDirt repo: https://github.com/Insane96/PathToDirt/issues/2
Stacktrace:
```java
java.lang.NullPointerException: Cannot invoke "net.minecraft.world.item.context.UseOnContext.m_43725_()" because "context" is null
	at net.minecraftforge.common.extensions.IForgeBlock.getToolModifiedState(IForgeBlock.java:680) ~[forge-1.18.2-40.1.57-universal.jar%23173!/:?] {re:mixin,re:classloading}
	at insane96mcp.pathtodirt.module.base.feature.BaseFeature.onRightClick(BaseFeature.java:61) ~[PathToDirt-1.4.2-mc1.18.2.jar%23138!/:1.4.2] {re:classloading}
	at net.minecraftforge.eventbus.ASMEventHandler_16_BaseFeature_onRightClick_BlockToolModificationEvent.invoke(.dynamic) ~[?:?] {}
	at net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:85) ~[eventbus-5.0.3.jar%232!/:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:302) ~[eventbus-5.0.3.jar%232!/:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:283) ~[eventbus-5.0.3.jar%232!/:?] {}
	at net.minecraftforge.event.ForgeEventFactory.onToolUse(ForgeEventFactory.java:382) ~[forge-1.18.2-40.1.57-universal.jar%23173!/:?] {re:classloading}
	at mrunknown404.slabgen.block.SlabGrass.m_6227_(SlabGrass.java:46) ~[SlabGen-1.18.2-1.4.2.jar%23144!/:1.4.2] {re:classloading}
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.m_60664_(BlockBehaviour.java:697) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,re:classloading,pl:mixin:APP:ferritecore.blockstatecache.mixin.json:BlockStateBaseMixin,pl:mixin:APP:snowrealmagic.mixins.json:BlockStateMixin,pl:mixin:A}
	at net.minecraft.client.multiplayer.MultiPlayerGameMode.m_105262_(MultiPlayerGameMode.java:296) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:mixins.ipnext.json:MixinScreenHandler,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.optifine.override.PlayerControllerOF.m_105262_(PlayerControllerOF.java:96) ~[OptiFine_1.18.2_HD_U_H7.jar%23174!/:?] {re:classloading}
	at net.minecraft.client.Minecraft.m_91277_(Minecraft.java:1571) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:balm.mixins.json:MinecraftMixin,pl:mixin:APP:bookshelf.common.mixins.json:client.AccessorMinecraft,pl:mixin:APP:mixin.dynamic_asset_generator.json:MinecraftMixin,pl:mixin:APP:mixins.ipnext.json:MixinMinecraftClient,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91279_(Minecraft.java:1851) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:balm.mixins.json:MinecraftMixin,pl:mixin:APP:bookshelf.common.mixins.json:client.AccessorMinecraft,pl:mixin:APP:mixin.dynamic_asset_generator.json:MinecraftMixin,pl:mixin:APP:mixins.ipnext.json:MixinMinecraftClient,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91398_(Minecraft.java:1668) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:balm.mixins.json:MinecraftMixin,pl:mixin:APP:bookshelf.common.mixins.json:client.AccessorMinecraft,pl:mixin:APP:mixin.dynamic_asset_generator.json:MinecraftMixin,pl:mixin:APP:mixins.ipnext.json:MixinMinecraftClient,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91383_(Minecraft.java:1019) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:balm.mixins.json:MinecraftMixin,pl:mixin:APP:bookshelf.common.mixins.json:client.AccessorMinecraft,pl:mixin:APP:mixin.dynamic_asset_generator.json:MinecraftMixin,pl:mixin:APP:mixins.ipnext.json:MixinMinecraftClient,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:663) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:balm.mixins.json:MinecraftMixin,pl:mixin:APP:bookshelf.common.mixins.json:client.AccessorMinecraft,pl:mixin:APP:mixin.dynamic_asset_generator.json:MinecraftMixin,pl:mixin:APP:mixins.ipnext.json:MixinMinecraftClient,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.client.main.Main.main(Main.java:205) ~[client-1.18.2-20220404.173914-srg.jar%23168!/:?] {re:classloading,pl:runtimedistcleaner:A}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?] {}
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?] {}
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?] {}
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?] {}
	at net.minecraftforge.fml.loading.targets.CommonClientLaunchHandler.lambda$launchService$0(CommonClientLaunchHandler.java:31) ~[fmlloader-1.18.2-40.1.57.jar%2316!/:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:37) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:106) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:77) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23) [modlauncher-9.1.3.jar%235!/:?] {}
	at cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:149) [bootstraplauncher-1.0.0.jar:?] {}
```

**Comment by MrUnknown404** [comment]
Thanks for the report, I'll look into it!

**Comment by Crystal-Spider** [comment]
I hope I can be of help.

Looking at the stacktrace, your code and Forge code this is what I found:
- The error happens because the context is null.
- The context is passed through `onToolUse` https://github.com/MrUnknown404/SlabGen/blob/e29b7c1995443e1a64b6aa76d8417af0a630fba4/SlabGen/src/main/java/mrunknown404/slabgen/block/SlabGrass.java#L46
- The `onToolUse` overload being used is the one that is removed in 1.19 and that creates the deprecated `BlockToolInteractEvent` with the context set to null:  
  - https://github.com/MinecraftForge/MinecraftForge/blob/1af16d70cc5d1a92cdbbeb5a44241bd36d89b7a4/src/main/java/net/minecraftforge/event/ForgeEventFactory.java#L379
  - https://github.com/MinecraftForge/MinecraftForge/blob/1af16d70cc5d1a92cdbbeb5a44241bd36d89b7a4/src/main/java/net/minecraftforge/event/world/BlockEvent.java#L434
  - https://github.com/MinecraftForge/MinecraftForge/blob/1af16d70cc5d1a92cdbbeb5a44241bd36d89b7a4/src/main/java/net/minecraftforge/event/world/BlockEvent.java#L495
- The solution I can think of would be to use the [other overload](https://github.com/MinecraftForge/MinecraftForge/blob/1af16d70cc5d1a92cdbbeb5a44241bd36d89b7a4/src/main/java/net/minecraftforge/event/ForgeEventFactory.java#L364) for `onToolUse` and get the context by doing `UseOnContext context = new UseOnContext(player, hand, ray);`, then get the `BlockState eventState` by doing `ForgeEventFactory.onToolUse(state, context, ToolActions.SHOVEL_FLATTEN, true);`.

With the solution proposed it fixes the game crash, I tested it both alone and with PathToDirt and I couldn't detect any bug or issue. I only tested the Grass slab.
However to make it work the Forge version must be updated to at least 40.1.0 (which anyway is the recommended version for 1.18.2).
Lastly, this solution doesn't make the grass/dirt/path slabs behave correctly with PathToDirt, meaning they still behave following the normal Vanilla behavior instead of the modded PathToDirt one, but at least it prevents the game crash and, in my opinion, it improves the code as it stops using a deprecated feature.

But let me know, I hope I was of any help.

**Comment by MrUnknown404** [comment]
Thanks for looking! I found out what was wrong right as you posted that lol.

> However to make it work the Forge version must be updated to at least 40.1.0

Yeah, This is basically the problem. The mod was made when Forge 1.18.2 came out so the newer `#onToolUse` didn't exist yet. I'll definitely update Forge.

> Lastly, this solution doesn't make the grass/dirt/path slabs behave correctly with PathToDirt, meaning they still behave following the normal Vanilla behavior instead of the modded PathToDirt one, but at least it prevents the game crash and, in my opinion, it improves the code as it stops using a deprecated feature.

This looks like it should be done in their config.

https://github.com/Insane96/PathToDirt/blob/2d9bdf852f824c6eeab2ad35272b9117b9f734ba/src/main/java/insane96mcp/pathtodirt/module/base/feature/BaseFeature.java#L66 ->
https://github.com/Insane96/PathToDirt/blob/2d9bdf852f824c6eeab2ad35272b9117b9f734ba/src/main/java/insane96mcp/pathtodirt/module/base/feature/BaseFeature.java#L31 ->
https://github.com/Insane96/PathToDirt/blob/2d9bdf852f824c6eeab2ad35272b9117b9f734ba/src/main/java/insane96mcp/pathtodirt/module/base/feature/BaseFeature.java#L52

**Comment by Crystal-Spider** [comment]
Yeah, you're right, adding these entries `"slabgen:path_slab,slabgen:dirt_slab", "slabgen:podzol_slab,slabgen:path_slab", "slabgen:coarse_dirt_slab,slabgen:path_slab"` to their config fixes the previously incorrect behavior.  
However I still get the crash for mycelium and dirt slabs, so I'm guessing the fix for the null context should be added there too.

**Comment by MrUnknown404** [comment]
> Yeah, you're right, adding these entries "slabgen:path_slab,slabgen:dirt_slab", "slabgen:podzol_slab,slabgen:path_slab", "slabgen:coarse_dirt_slab,slabgen:path_slab" to their config fixes the previously incorrect behavior.

Good to know!
> However I still get the crash for mycelium and dirt slabs, so I'm guessing the fix for the null context should be added there too.

Yeah, they're all using the same method. I'll have an update out soon to fix it.

**Comment by Crystal-Spider** [comment]
Okay then, thanks for the fast replies and the effort!  
I'll look forward to the fix!

**Comment by MrUnknown404** [comment]
Fixed in 1.5.0
